### ISSUE TITLE
TensorFlow Jupyter Notebook images 1.9 and above in gcr.io cannot see GPUs

### ISSUE THREAD

**Issue by mikereeves** [kubeflow/kubeflow#1828]
### What works: 1.8.0

gcr.io/kubeflow-images-public/tensorflow-1.8.0-notebook-gpu:v0.3.0 

<img width="1148" alt="screen shot 2018-10-19 at 8 17 43 am" src="https://user-images.githubusercontent.com/6353162/47217814-a48be280-d377-11e8-971d-76ed15368dee.png">

### What does not work: 1.9.0 and 1.10.1

gcr.io/kubeflow-images-public/tensorflow-1.9.0-notebook-gpu:v0.3.0
gcr.io/kubeflow-images-public/tensorflow-1.10.1-notebook-gpu:v0.3.0

<img width="1147" alt="screen shot 2018-10-19 at 8 54 29 am" src="https://user-images.githubusercontent.com/6353162/47219368-9f7d6200-d37c-11e8-8b64-5c826a6f1aec.png">

### Checking scheduling on GPU node and limits

`kubectl get pods --output=wide

NAME                                         READY     STATUS      RESTARTS   AGE       IP           NODE
ambassador-7fb86f6bc5-dsmdr                  3/3       Running     0          2h        10.20.0.11   gke-dsp-2-default-pool-2e332159-dt8n
ambassador-7fb86f6bc5-qs2z8                  3/3       Running     0          2h        10.20.0.10   gke-dsp-2-default-pool-2e332159-dt8n
ambassador-7fb86f6bc5-rfsts                  3/3       Running     0          2h        10.20.1.12   gke-dsp-2-default-pool-2e332159-k2cv
argo-ui-7b6585d85d-4tk55                     1/1       Running     0          2h        10.20.1.16   gke-dsp-2-default-pool-2e332159-k2cv
centraldashboard-f8d7d97fb-lwh4g             1/1       Running     0          2h        10.20.1.14   gke-dsp-2-default-pool-2e332159-k2cv
cert-manager-798d77c76-tkvqz                 1/1       Running     0          2h        10.20.1.23   gke-dsp-2-default-pool-2e332159-k2cv
cloud-endpoints-controller-b55c586b6-ntslm   1/1       Running     0          2h        10.20.0.17   gke-dsp-2-default-pool-2e332159-dt8n
envoy-79ff8d86b-2ghxx                        2/2       Running     2          2h        10.20.1.25   gke-dsp-2-default-pool-2e332159-k2cv
envoy-79ff8d86b-hn4fp                        2/2       Running     2          2h        10.20.0.22   gke-dsp-2-default-pool-2e332159-dt8n
envoy-79ff8d86b-lcjks                        2/2       Running     2          2h        10.20.1.24   gke-dsp-2-default-pool-2e332159-k2cv
iap-enabler-6565b855d6-qcfkm                 1/1       Running     0          2h        10.20.0.21   gke-dsp-2-default-pool-2e332159-dt8n
ingress-bootstrap-5wqrr                      0/1       Completed   0          2h        10.20.0.19   gke-dsp-2-default-pool-2e332159-dt8n
**jupyter-mike-2reeves-40xxxxxxxx-2enet        1/1       Running     0          2m        10.20.2.14   gke-dsp-2-gpu-pool-165972e2-ks2n**
kube-metacontroller-6d567f9fb4-gc26z         1/1       Running     0          2h        10.20.1.22   gke-dsp-2-default-pool-2e332159-k2cv
modeldb-backend-69dfc464df-5ch5l             1/1       Running     0          2h        10.20.1.21   gke-dsp-2-default-pool-2e332159-k2cv
modeldb-db-6cf5bb764-5hml9                   1/1       Running     0          2h        10.20.1.17   gke-dsp-2-default-pool-2e332159-k2cv
modeldb-frontend-795bcf6df9-ls7m8            1/1       Running     0          2h        10.20.0.14   gke-dsp-2-default-pool-2e332159-dt8n
pytorch-operator-76f89745c-stcrx             1/1       Running     0          2h        10.20.0.23   gke-dsp-2-default-pool-2e332159-dt8n
spartakus-volunteer-5b7bb957c6-rt2qn         1/1       Running     0          2h        10.20.0.16   gke-dsp-2-default-pool-2e332159-dt8n
studyjob-controller-68f5948984-2ssgm         1/1       Running     0          2h        10.20.0.15   gke-dsp-2-default-pool-2e332159-dt8n
tf-hub-0                                     1/1       Running     0          1h        10.20.0.24   gke-dsp-2-default-pool-2e332159-dt8n
tf-job-dashboard-7cddcdf9c4-nkd75            1/1       Running     0          2h        10.20.1.15   gke-dsp-2-default-pool-2e332159-k2cv
tf-job-operator-v1alpha2-6566f45db-tq4pj     1/1       Running     0          2h        10.20.0.12   gke-dsp-2-default-pool-2e332159-dt8n
vizier-core-6d8c9d7bf7-vsc4x                 1/1       Running     2          2h        10.20.1.19   gke-dsp-2-default-pool-2e332159-k2cv
vizier-db-cc59bc8bd-ktxhs                    1/1       Running     0          2h        10.20.0.18   gke-dsp-2-default-pool-2e332159-dt8n
vizier-suggestion-grid-76fb6b49c7-9mbzb      1/1       Running     0          2h        10.20.1.18   gke-dsp-2-default-pool-2e332159-k2cv
vizier-suggestion-random-c5c64dfc9-m85bz     1/1       Running     0          2h        10.20.1.20   gke-dsp-2-default-pool-2e332159-k2cv
whoami-app-b7fb9f875-wsmdj                   1/1       Running     0          2h        10.20.0.20   gke-dsp-2-default-pool-2e332159-dt8n
workflow-controller-59c7967f59-cg2kf         1/1       Running     0          2h        10.20.0.13   gke-dsp-2-default-pool-2e332159-dt8n`

## Limits check

`kubectl -n kubeflow describe pod jupyter-mike-2reeves-40xxxxxxxx-2enet
Name:         jupyter-mike-2reeves-40xxxxxxxx-2enet
Namespace:    kubeflow
Node:         gke-dsp-2-gpu-pool-165972e2-ks2n/10.142.0.4
Start Time:   Fri, 19 Oct 2018 08:53:13 -0400
Labels:       app=jupyterhub
              component=singleuser-server
              heritage=jupyterhub
Annotations:  hub.jupyter.org/username=accounts.google.com:mike-2reeves-40xxxxxxxx-2enet
Status:       Running
IP:           10.20.2.14
Containers:
  notebook:
    Container ID:  docker://53ecaaf0dc32f21226934ac031b9273e31ede7c99d5a602a7fbdfb252428dc0f
    Image:         gcr.io/kubeflow-images-public/tensorflow-1.10.1-notebook-gpu:v0.3.0
    Image ID:      docker-pullable://gcr.io/kubeflow-images-public/tensorflow-1.10.1-notebook-gpu@sha256:24a32e6a533a6e7d46a649cbff247ebd86dee078d37489035d9bc440a9e2487f
    Port:          8888/TCP
    Host Port:     0/TCP
    Args:
      start-singleuser.sh
      --ip="0.0.0.0"
      --port=8888
      --allow-root
    State:          Running
      Started:      Fri, 19 Oct 2018 08:53:23 -0400
    Ready:          True
    Restart Count:  0
    **Limits:
      nvidia.com/gpu:  1
    Requests:
      cpu:             1
      memory:          2Gi
      nvidia.com/gpu:  1**
      
      - - - - - -
      
      `

## nvidia-smi 

`      kubectl exec -it jupyter-mike-2reeves-40xxxxxxxx-2enet -- /bin/bash
      jovyan@jupyter-mike-2reeves-40xxxxxxxx-2enet:~$ nvidia-smi
      Fri Oct 19 12:57:39 2018
      +-----------------------------------------------------------------------------+
      | NVIDIA-SMI 396.26                 Driver Version: 396.26                    |
      |-------------------------------+----------------------+----------------------+
      | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
      | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
      |===============================+======================+======================|
      |   0  Tesla K80           Off  | 00000000:00:04.0 Off |                    0 |
      | N/A   30C    P8    26W / 149W |      0MiB / 11441MiB |      0%      Default |
      +-------------------------------+----------------------+----------------------+

      +-----------------------------------------------------------------------------+
      | Processes:                                                       GPU Memory |
      |  GPU       PID   Type   Process name                             Usage      |
      |=============================================================================|
      |  No running processes found                                                 |
      +-----------------------------------------------------------------------------+`

## Running TF version 1.10.1 directory on the container shows the GPU device(s)

`
      jovyan@jupyter-mike-2reeves-40xxxxxxxx-2enet:~$ python
      Python 3.6.4 |Anaconda, Inc.| (default, Jan 16 2018, 18:10:19)
      [GCC 7.2.0] on linux
      Type "help", "copyright", "credits" or "license" for more information.
      >>> import tensorflow as tf
      >>> print(tf.__version__)
      **1.10.1**
      >>> from tensorflow.python.client import device_lib
      >>> def get_available_devices():
      ...     local_device_protos = device_lib.list_local_devices()
      ...     return [x.name for x in local_device_protos]
      ...
      >>> print(get_available_devices())
      2018-10-19 12:58:51.296353: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
      2018-10-19 12:58:51.393328: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:897] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
      2018-10-19 12:58:51.393906: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 0 with properties:
      name: Tesla K80 major: 3 minor: 7 memoryClockRate(GHz): 0.8235
      pciBusID: 0000:00:04.0
      totalMemory: 11.17GiB freeMemory: 11.10GiB
      2018-10-19 12:58:51.393966: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0
      2018-10-19 12:58:51.714013: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
      2018-10-19 12:58:51.714074: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0
      2018-10-19 12:58:51.714110: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N
      2018-10-19 12:58:51.714391: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/device:GPU:0 with 10758 MB memory) -> physical GPU (device: 0, name: Tesla K80, pci bus id: 0000:00:04.0, compute capability: 3.7)
      **['/device:CPU:0', '/device:GPU:0']**
      >>>

`

**Comment by jlewi** [comment]
I tried: gcr.io/kubeflow-ci/tensorflow-1.10.1-notebook-gpu:v-base-81622af-pull-da32edf-pr-1815-3948

And when I printed the devices using TF I saw GPUs.
So I think this should be fixed in the 0.3.1 images that we should cut soon.

See #1718

**Comment by jlewi** [comment]
Here's the snippet I ran to check TF could see the devices.

```
from tensorflow.python.client import device_lib

def get_available_devices():
   local_device_protos = device_lib.list_local_devices()
   return [x.name for x in local_device_protos]

print(get_available_devices())
```

**Comment by jlewi** [comment]
@mikereeves If you want to give it a try here's a new image:
docker pull gcr.io/kubeflow-images-public/tensorflow-1.8.0-notebook-gpu:v-base-b321075-822

**Comment by mikereeves** [comment]
@jlewi Thanks. Will give it a shot!

**Comment by jlewi** [comment]
Other TF versions are available as well just change the TF version
e.g
docker pull gcr.io/kubeflow-images-public/tensorflow-1.9.0-notebook-gpu:v-base-b321075-822

**Comment by mikereeves** [comment]
@jlewi http://gcr.io/kubeflow-ci/tensorflow-1.10.1-notebook-gpu:v-base-81622af-pull-da32edf-pr-1815-3948 and http://gcr.io/kubeflow-images-public/tensorflow-1.9.0-notebook-gpu:v-base-b321075-822 worked. Good to go for now!...look forward to 0.3.1. Thanks again.

**Comment by jlewi** [comment]
Thanks for confirming.
